src/components/lib/shm_bm/shm_bm.c

```diff
@@ -0,0 +1,193 @@
+#define SHM_BM_NOBJ(shm) (((struct shm_bm_header *) shm)->nobj)
+
+
+static inline void
+shm_bm_set_contig(word_t *bm, int offset)
+{
+	int i, n, ind;
+
+	ind = offset / SHM_BM_BITMAP_BLOCK;
+	offset %= SHM_BM_BITMAP_BLOCK;
+
+	for (i = 0; i < ind; i++)
+		bm[i] = ~0x0ul;
+
+	n = SHM_BM_BITMAP_BLOCK - offset;
+	bm[ind] = ~((1ul << n) - 1); // set most sig n bits of bm[ind]
+}
+
+static int
+shm_bm_clz(word_t *bm, void *stop, int *index, int *offset)
+{
+	int cnt, lz = 0;
+
+	while (*bm == 0) {
+		if ((void *)bm == stop) return -1;
+		bm++;
+		(*index)++;
+		lz += SHM_BM_BITMAP_BLOCK;
+	}
+
+	if ((void *)bm >= stop) return -1;	
+	cnt = __builtin_clzl(*bm);
+	lz += cnt;
+	*offset = SHM_BM_BITMAP_BLOCK - cnt - 1;
+	return lz;
+
+}
+
+cbuf_t   
+shm_bm_create(shm_bm_t *shm, size_t objsz, size_t allocsz)
+{
+	struct shm_bm_header *header;
+	int                   nobj;
+	size_t                bitmap_sz, refcnt_sz, data_sz;
+	size_t                alloc;
+	cbuf_t                id;
+
+	if (allocsz < objsz) return 0;
+
+	nobj = allocsz / objsz;
+	bitmap_sz = SHM_BM_BITS_TO_WORDS(nobj) * sizeof (word_t);
+	refcnt_sz = nobj;
+	data_sz   = nobj * objsz;
+
+	alloc = sizeof (struct shm_bm_header) + bitmap_sz + refcnt_sz + data_sz;
+	if (alloc > SHM_BM_ALLOC_ALIGNMENT) return 0;
+
+	id  = memmgr_shared_page_allocn_aligned(round_up_to_page(alloc)/PAGE_SIZE, SHM_BM_ALLOC_ALIGNMENT, (vaddr_t *) shm);
+
+	if (id == 0) return 0;
+	memset((void *) *shm, 0, round_up_to_page(alloc));
+
+	// metadata
+	header = (struct shm_bm_header *) *shm;
+	header->objsz       = objsz;
+	header->nobj        = nobj;
+	header->bitm_offset = sizeof (struct shm_bm_header); 
+	header->refc_offset = sizeof (struct shm_bm_header) + bitmap_sz; 
+	header->data_offset = sizeof (struct shm_bm_header) + bitmap_sz + refcnt_sz; 
+
+	// set nobj bits to free
+	shm_bm_set_contig(SHM_BM_BITM(*shm), nobj);
+	return id;
+}
+
+shm_bm_t 
+shm_bm_map(cbuf_t id)
+{
+	shm_bm_t shm;
+	if (memmgr_shared_page_map_aligned(id, SHM_BM_ALLOC_ALIGNMENT, (vaddr_t *) &shm) == 0) return 0;
+	return shm;
+}
+
+void *
+shm_bm_obj_alloc(shm_bm_t shm, shm_bufid_t *id)
+{
+	int     freebit = 0, idx = 0, off = -1;
+	word_t  word_old; 
+	word_t *bm;
+
+	// find a free space. could be preempted
+	bm = SHM_BM_BITM(shm);
+	do {
+		freebit = shm_bm_clz(bm, SHM_BM_REFC(shm), &idx, &off);
+		if (freebit == -1) return 0;
+		word_old = bm[idx];
+	} while (!cos_cas(bm + idx, word_old, word_old & ~(1ul << off)));
+
+	cos_faab(SHM_BM_REFC(shm) + freebit, 1);
+
+	*id = (shm_bufid_t) freebit;
+	return SHM_BM_DATA(shm) + (freebit * SHM_BM_SIZE(shm));
+}
+
+void *   
+shm_bm_obj_use(shm_bm_t shm, shm_bufid_t id)
+{
+	if (id >= SHM_BM_NOBJ(shm)) return 0;
+
+	// obj has not been allocated
+	if ((SHM_BM_REFC(shm) + id) == 0) return 0;
+
+	cos_faab(SHM_BM_REFC(shm) + id, 1);
+	return SHM_BM_DATA(shm) + (id * SHM_BM_SIZE(shm));
+}
+
+void *   
+shm_bm_obj_take(shm_bm_t shm, shm_bufid_t id)
```

Comment:
Kinda confused as I figured there would be some modification to the structure to refcnt. I suppose this might be the optimization that for passing ownership.

Reply:
Yes, this is documented in the header file; im trying to think of a better name than "take"